### ISSUE TITLE
Parallelism within the package tree

### ISSUE THREAD

**Issue by bengry** [lerna/lerna#1938]
It's not clear how lerna parallelism works in regards to package dependency and the build order.
For example, if I have the follow dependency tree (via `devDependencies`/`dependencies`):
```
packages/
  +--package1
  +   +--package3
  +   +--package2
  +--package2
  +   +--package4
  +--package3
  +   +--package4
  +--package4

```
When I run `lerna run build`, obviously `package4` is built first, and `package1` is built last, but then it's not clear if `package2` and `package4` are build in parallel, or one after the other (even though they have no relation to each other.

## Expected Behavior
Packages should be built in parallel as much as possible.

## Current Behavior
Not sure. The docs aren't clear on this.

## Context
- How has this issue affected you? What are you trying to accomplish?
  - Faster build times

## Your Environment
| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.11.1 |
| `npm --version`  | 6.7.0 |
| `node --version` | 10.12.0 |

| OS | Version |
| --- | --- |
| macOS Mohave | 10.14.2 |

**Comment by evocateur** [comment]
Lerna [batches packages](https://github.com/lerna/lerna/blob/master/utils/batch-packages/batch-packages.js) into maximally-concurrent groups, respecting topological order by default. `npx lerna ls --toposort` shows you the (flattened) order derived from this.

Within each batch, the packages are generally sorted lexically (alphabetical by path). `--concurrency` determines how many packages at a time, per batch, are "in flight". The order of resolution within a batch is non-deterministic (`Promise`s, yay). All packages in a batch will complete before the next batch is started ([run-parallel-batches](https://github.com/lerna/lerna/blob/master/utils/run-parallel-batches/run-parallel-batches.js)).

It's not documented very deeply because it's an implementation detail. All users need to know is that most operations are run as concurrently as possible, respecting topological order by default.

**Comment by bengry** [comment]
Great, that what I was looking to hear. Thanks!
Is there a way to show the full dependency tree, as determined by lerna? As
an actual tree structure, and not a flat list. Either from the cli or an
API we can build a small script on top of.

**Comment by evocateur** [comment]
Not directly, though both [`lerna exec`](https://github.com/lerna/lerna/tree/master/commands/exec#readme) and [`lerna run`](https://github.com/lerna/lerna/tree/master/commands/run#readme) will execute their arguments with maximal topological concurrency.

Programmatically, all the bits are available as separate packages, so it wouldn't be too hard to retrieve the same topologically batched package list that Lerna uses internally:
```js
const { getPackages } = require('@lerna/project');
const batchPackages = require('@lerna/batch-packages');

(async () {
  const cwd = process.cwd();
  const pkgs = await getPackages(cwd);
  const batches = batchPackages(pkgs);

  for (const batch of batches) {
    // all packages in a batch are safe to operate on concurrently
    for (const pkg of batch) {
      console.log(pkg.name);
      console.log(pkg.location); // absolute path to leaf directory
    }
  }
}());

**Comment by wbern** [comment]
How could this work if you run scripts in watch mode? Any way to signal to lerna when a script is "compiled" despite that it keeps running?

**Comment by bengry** [comment]
@evocateur Thanks for the detailed response. Exactly what I was looking for doing.
Closing the issue.

**Comment by evocateur** [comment]
@wbern That's a separate issue, which I believe already exists (possibly in triplicate).
